### ISSUE TITLE
Terminology

### ISSUE THREAD

**Issue by fimbault** [ietf-wg-gnap/gnap-core-protocol#29]
People may misunderstand a term which has a specific meaning in the context of an authorization framework. The objective of this issue is to clarify the terms we use to avoid confusion.

In the current version of the GNAP spec (01), we have : 

* parties in the GNAP protocol perform actions under different roles; roles are defined by the actions taken and the expectations leveraged on the role by the overall protocol ; elements are acted upon by the roles throughout the process
* In section 1.2 (roles): Authorization Server (AS), Resource Client (RC, aka "client"), Resource Server (RS, aka "API"), Resource Owner (RO), Requesting Party (RQ, aka "user")
* In section 2.3: new notion of RC instance. This was added to allow a mecanism analogous to OAuth 2's "client_id" parameter, but it also relates to issues discussed within the OAuth WG (e.g. credential client, as an addition to public and confidential client types). It brings additional potential (attestation framework, display name, device posture, etc.)
* In section 1.3 (elements): Access Token, Grant, Cryptographic Key, Resource, Subject Information
* We also find definitions in the request/responses, e.g. resources, interaction, claims (see section 2), assertions (see 2.2)

a) are we fine with those names, abbreviations and definitions? 
- Define access token without using "credential"
- Subject information should be about the user, not the RO
- Grant as both a noun (what is being requested) and a verb?

note : we should avoid acronyms which are already widely used in our software domain (such as RP = relying party)

b) are there missing terms? 
- what is a refreshed credential? (step 7 in section 1.3.3)
- (maybe) device posture/attestation if that's something we develop further
- check all the terms in the request/responses

c) do we need a section dedicated to terminology (to define all the terms), or do we define at first use?
In ISO for instance, terms are defined at the beginning of the spec (nice for lookup). It would be interesting to provide sources to  pre-existing terminology (in our case, there's for instance https://canada-ca.github.io/PCTF-CCP/PCTF.html, GDPR/CCPA/etc.). We can then have notes (to explain what differs from a commonly accepted definition, in our specific context) and examples.
I would be in favor of doing the same. 

When commenting, please provide why you think the term and/or definition should be modified, and an alternative proposal. The objective is to prepare a PR which consolidate the changes related to terminology.

**Comment by yaronf** [comment]
+1 to a separate Terminology section, for the most important 10-20 terms.

**Comment by dickhardt** [comment]
+1 to a separate Terminology section. A PR for changing the terminology scattered across the document will be hard to follow.

How about we create a wiki with the current terminology, and then we the WG can work with that. Then the editors can do a PR that fixes up the draft?

**Comment by fimbault** [comment]
Good idea. Available at https://github.com/ietf-wg-gnap/gnap-core-protocol/wiki/Terminology

**Comment by jricher** [comment]
I think we should move away from "client" and toward explicitly talking about "client instance" throughout the document, going so far as replacing "Resource Client / RC" with "Client Instance / CI" throughout.

**Comment by TomCJones** [comment]
two issues that i care about:
1. RC or CI or whatever, must have an ID or it does not exist in any meaningful sense. If you can't identify something, you cannot talk about it.
2. "Subject information should be about the user, not the RO" is ontologically false in the typical case were data about the subject, which is ontologically correct, the the "owner" (in the sense meant by the std, as opposed to real-life), is the subject of the data.
3. Owner is problematic for the reason listed in 2, but i can live with it, provided sometime we can say that data about the RO can be, in fact, legally owned by the RS even tho it is about the RO. This is a problem inherited from the GDPR so i suppose we are stuck with.

**Comment by francis-pouatcha** [comment]
Instead of having a wiki page I suggest we create a directory and have foreach term:
- A file that carries the agreed on version of the description of that term.
- A issue to document movement that file
- while keeping detailed discussion on the mailing list

**Comment by fimbault** [comment]
@francis-pouatcha Keeping the detailed discussion on the mailing list and on the git issues will be a nightmare. I guess we should address that at the next IETF meeting. We need to keep the process simple, and yet provide full transparency and auditability on why we landed on a specific definition (which is hard to do through the mailing list, how would you keep that in sync with the issue?).
Even if I hope we can quickly converge on a majority of the terms, I expect that there will be updates till the final publication of the spec.

So I've thought about it a bit more about the process related to terminology, and here's what I propose : 

* this issue #29 will track the global changes related to terminology within the spec. In any case, the only file that caries the agreed version (the source of truth) should be the spec itself, and here we'll have a PR process (and at some point a vote on the mailing list). All movements shall be tracked here (possibly linking more specific issues).

* the wiki would be used as a living document (yet versioned) that provides a global list of terms. In case we all converge quickly on a term, there's not even the need to open an issue. Else a dedicated issue will enable to track the advances.

**Comment by fimbault** [comment]
@TomCJones thanks for your input

1. You can identify an instance, see section 2.3.1 (instance_id). This makes Justin's proposal interesting, because it clarifies what we mean by "client". 

>we should move away from "client" and toward explicitly talking about "client instance" throughout the document

2. Subject information : RO or user? This doesn't seem so clear. Dick provided an example to clarify: "For example, in an enterprise use case where resource access and identity claims are being requested, the RO is not the user, but could be an administrator. The "subject information" is going to be about the user of the client, not the administrator." Maybe that's where examples could come handy, because we actually have 2 different cases : one for personal accounts (user = RO in most cases), one for corporate use (user != RO).

3.  And there's the (somewhat) grey area where data storage and processing has been delegated to a third party, currently being adressed by regulators (for instance GDPR). That's typically where we could insert a note ("current regulation such as GDPR refer to data owners, which....") that helps contextualize our definition in a broader context.

**Comment by fimbault** [comment]
Additional input on the RO, when reviewing Francis feedback on v01 : 
"Authorizes the request? Does it authorize the request or the access to a resource?"

We should change the wording to clarify.

**Comment by jricher** [comment]
In my opinion a wiki isn't the best way to arrive at consensus or have discussion for things. A wiki is a good tool for capturing consensus and gathering new terms, like we did with the use cases: https://github.com/ietf-wg-gnap/general/wiki/Use-Cases

Also, any wikis and related tools should probably be on the `ietf-wg-gnap/general` repository and not this one.

**Comment by fimbault** [comment]
ok let's review this during the next IETF meeting!

**Comment by fimbault** [comment]
Following the meeting, I reorganized the [terminology related wiki](https://github.com/ietf-wg-gnap/gnap-core-protocol/wiki/Terminology)

Next step is for concrete proposals (and we'll need to make sure we have something consistent before the PR).

**Comment by yaronf** [comment]
Quoting from the wiki:
> Access Token: A credential representing a set of access rights delegated to the RC. The access token is created by the AS, consumed and verified by the RS, and issued to and carried by the RC. The contents and format of the access token are opaque to the RC.

Currently access tokens are doing double duty, both enabling the RC to access a resource (its main job, described above) but also internal to the RC-AS protocol. The text doesn't do justice to this other role.

**Comment by fimbault** [comment]
@yaronf thanks for spotting that, depends on resolving issue #67. 
* Should access token be mandatory?
* Should access token be removed?
(and maybe as you suggested, rename the structure inside the json "access_token" by "access")

If we decide that we indeed use access token for both RS and AS, it should be reflected in the definition

**Comment by jricher** [comment]
I agree that the access token's definition should be expanded. Since the AS is now presenting an API, we could say:

> Access Token: A credential representing a set of access rights to an API delegated to the RC. The APIs protected by an access token are hosted by the RS, with the exception of the delegation API described in this specification which is provided by the AS. The access token is created by the AS, and the token is issued to and carried by the RC. The contents and format of the access token are opaque to the RC.

**Comment by yaronf** [comment]
I think the 2nd sentence is too detailed for a terminology section. I would suggest:
> Access Token: A credential representing a set of access rights to an API delegated to the RC. The APIs protected by an access token are typically hosted by the RS, but may also be hosted by the AS. The access token is created by the AS, and the token is issued to and carried by the RC. The contents and format of the access token are opaque to the RC.

**Comment by fimbault** [comment]
Closing #32 and #8 as duplicates

§1.2 Roles: Editor's note:

The names for the roles are an area of ongoing
discussion within the working group, as is the appropriate precision
of what activities and expectations a particular role covers. In particular,
the AS might be formally decomposed into delegation components, that the
client talks to, and interaction components, that the user talks to. Several
alternative names have been proposed for different roles and components,
including:

Grant Server (for Authorization Server)
Grant Client (for Resource Client)
Operator (for Requesting Party)

-------------------------------------------------------------------------------------------------------------------
Re: the "Elements" section.

Yaron: Can we call it something more specific, maybe "holder identity key"?

Justin: This isn’t just about a user’s key — this is primarily a key for the requesting client software, but it gets used in a number of ways. Keys can be bound to client instances and to access tokens, and they can be used to identify various parties in the system (including users).

It was brought up in the design team meetings that we might not even need to define this term, since it’s understood in the wider security space, and so I’m happy to either take it out or expand it as required.

Yaron: I tend to agree we should remove it.

**Comment by dickhardt** [comment]
The access token has a well understood meaning as currently defined in terminology. Redefining the term will confuse people. 

I suggest changing "A credential representing .." to "A string representing .."  => the reader is not having to understand what a credential is, and so far access tokens are only strings.

**Comment by fimbault** [comment]
@dickhardt I support removing credential in the definition. Indeed there could be various encodings but in the end it is always a string. Since we have "representing..." the definition conveys the difference with any other strings, so it is distinct enough.

**Comment by jricher** [comment]
The goal is not to redefine an access token, but to be specific about how it fits in this protocol. In fact the "credential terminology" was imported from OAuth 2's own definition. I'm fine with dropping it from the definition here, though.

**Comment by fimbault** [comment]
Closing issue with merge of #155 
New changes (new terms, change to an existing definition) will be managed through dedicated issues now that we have a solid basis.
